### PR TITLE
fix: fix percentage label consistency across all data-viz components

### LINKED ISSUE #5474: fix: ensure sub-1% donut chart slices are always visible in ResponsiveDonutWithInnerPercent
## Problem

nivo's `ResponsivePie` has no `minValue` prop. When a slice's value is less than ~0.5% of the total, it renders as an invisible arc — the slice simply disappears even though the data is non-zero. This affects every consumer of `ResponsiveDonutWithInnerPercent`.

This was discovered while building the Projects list summary cards (#5473), where Research budget (~$3.5B) completely swallows the Admin & Support slice (~$301K, ≈0.008% of total).

## Current workaround

`ProjectTypeSummaryCard` works around this locally with an `applyMinimumArcValue` helper that floors any non-zero value below 1% of the total up to that threshold before passing data to the chart. This keeps the real values in the legend while only adjusting chart geometry.

```js
const applyMinimumArcValue = (items, total) => {
    if (total === 0) return items;
    const minValue = total * 0.01;
    return items.map((item) => ({
        ...item,
        value: item.value > 0 && item.value < minValue ? minValue : item.value
    }));
};
```

## Proposed fix

Move this logic into `ResponsiveDonutWithInnerPercent` directly so all current and future consumers get the behavior automatically. The component should apply the floor to its `data` prop internally before passing to `ResponsivePie`.

Consumers that need to display real values in their legends (as `ProjectTypeSummaryCard` does) should keep separate `legendData` (real values) and `chartData` (floored values) — the fix inside the shared component only affects what nivo renders, not what the parent passes in.

## Affected consumers

- `AgreementSpendingSummaryCard` — agreement types (Contract, Grant, IAA, Direct Obligation)
- `BLIStatusSummaryCard` — BLI statuses (Draft, Planned, Executing, Obligated, Available)
- `CANSummaryCards` — CAN budget breakdown
- `ProjectTypeSummaryCard` — already has local workaround (can be removed after global fix)

## Acceptance criteria

1. `ResponsiveDonutWithInnerPercent` applies a 1%-of-total minimum arc value to any non-zero slice internally
2. All existing consumers render correctly — no visual regression on balanced data
3. A consumer with one dominant value (e.g. 99.99% / 0.01% split) shows a visible thin slice for the minority category
4. The local `applyMinimumArcValue` workaround in `ProjectTypeSummaryCard` is removed
5. Unit tests updated for `ResponsiveDonutWithInnerPercent`

## References

- Introduced workaround: #5473

### LINKED ISSUE #5513: story: fix percentage label consistency across all data-viz components
## Background

While fixing a contradictory `100% + <1%` label bug in `ProjectTypeSummaryCard` (#5511), a codebase-wide audit revealed that the same class of rounding and display-consistency bugs affects multiple data visualization components. A related arc-visibility issue is already tracked in #5474.

This story tracks the full remediation across all affected components.

## Problem

Most data-viz components that display percentage labels compute them **independently per item** using `calculatePercent` (or inline `Math.round`/`Math.floor`). There is no cross-item normalization step, which causes two classes of bug:

1. **Rounding contradiction** — A dominant item at ~99.6% rounds to `100%` while a co-existing non-zero item rounds to `0%` and shows `<1%`. Displaying `100% + <1%` simultaneously is logically contradictory.
2. **Sum drift** — Independent rounding can cause percents to sum to 99% or 101% (e.g. three equal items: `33% + 33% + 33% = 99%`).

A third variant — arc visibility — is tracked separately in #5474.

### HIGH risk — independent per-item `calculatePercent`, visible `%` labels, no guards

| Component | Slices | Notes |
|---|---|---|
| `BLIStatusSummaryCard` | 4 donut slices | No guards at all |
| `AgreementSpendingSummaryCard` | 4 donut slices | Test even demonstrates `33+33+33+0 = 99%` without flagging it |
| `PortfolioFunding` | 2 items | Semantically complementary values computed independently instead of using `x` / `100-x` |
| `AgreementSpendingCards.helpers.js` | N segments in loop | Feeds `HorizontalStackedBar`; no guards |

### MEDIUM risk — partial guards only

| Component | Gap |
|---|---|
| `PortfolioSummaryCards` / `PortfolioLegend` | Has `<1%` guard but **no `>99%` guard** — dominant portfolio can still show `100%` alongside `<1%` |
| `HorizontalStackedBar` | CSS min-width floor for tiny bars, but sr-only and aria-label text still speaks the raw unguarded percent — a11y concern |

### Critical isolated bug

| Component | Issue |
|---|---|
| `PortfolioFundingByCAN` | Uses `Math.floor` (not `Math.round`) and bakes `%` into the string value as `"33%"` — systematically undercounts, incompatible with every other component's numeric convention, and invisible to any future shared utility |

### Safe by design (no action needed)

`BudgetCard`, `BigBudgetCard`, `ReceivedFundingCard`, `CanCard` — all use the two-item `[x, 100-x]` arithmetic complement pattern, immune to this class of bug.

### Reference implementation

`ProjectTypeSummaryCard` is the **fully hardened** reference:
- `computeDisplayPercents` — processes all items together, applies `>99%` cap when dominant item would round to `100%` alongside non-zero peers, returns `<1%` for sub-1% non-zero values
- `applyMinimumArcValue` — floors tiny chart arcs to ≥1% of total with redistribution (arc visibility, per #5474)
- Separates legend data (real values) from chart data (padded values)

## Acceptance Criteria

- [x] Extract `computeDisplayPercents` from `ProjectTypeSummaryCard` into `src/helpers/utils.js` as a shared utility
- [x] `BLIStatusSummaryCard` — replace independent `calculatePercent` calls with shared `computeDisplayPercents`
- [x] `AgreementSpendingSummaryCard` — replace independent `calculatePercent` calls with shared `computeDisplayPercents`
- [x] `PortfolioFunding` — use `x` / `100-x` complement pattern or `computeDisplayPercents`
- [x] `AgreementSpendingCards.helpers.js` — apply `computeDisplayPercents` to segment loop
- [x] `PortfolioLegend` — add `>99%` guard to complement existing `<1%` guard
- [x] `PortfolioFundingByCAN` — replace `Math.floor` with `Math.round`, use numeric convention (no baked-in `%` string), apply cross-item guards
- [x] `HorizontalStackedBar` — ensure sr-only and aria-label text reflects the guarded display percent, not the raw value
- [x] `ResponsiveDonutWithInnerPercent` arc visibility fix (closes #5474)
- [x] All existing unit tests pass; new tests added for edge cases (dominant item, equal splits, single item)

## References

- #5474 — arc visibility fix (child)
- #5511 — `ProjectTypeSummaryCard` label contradiction fix (child, completed)

### PR DESCRIPTION
**Pull request by fpigeonjr**
## What changed

This PR addresses a codebase-wide class of rounding and display-consistency bugs affecting all data-viz components, and fully closes the arc-visibility issue in #5474.

**Root cause:** Most data-viz components computed percentage labels independently per item using `calculatePercent`. With no cross-item normalisation, a dominant item at ~99.6% would round to `100%` while a co-existing non-zero item would round to `0%` and show `<1%` — a logical contradiction. Three equal items would show `33% + 33% + 33% = 99%` (sum drift).

**Fix:** Three shared utilities extracted into `src/helpers/utils.js`:
- `computeDisplayPercent` — per-item guard (`<1` for sub-1% non-zero)
- `computeDisplayPercents` — cross-item normaliser (`>99` cap when dominant item alongside non-zero peers)
- `applyMinimumArcValue` — floors tiny donut arcs to ≥1% of total (now applied inside `ResponsiveDonutWithInnerPercent` for all consumers automatically)

**Components fixed:**
| Component | Fix |
|---|---|
| `BLIStatusSummaryCard` | `computeDisplayPercents` replaces 4× independent `calculatePercent` |
| `AgreementSpendingSummaryCard` | Same treatment |
| `AgreementSpendingCards.helpers.js` | `computeDisplayPercents` applied to full segment array after build |
| `PortfolioFunding` | `computeDisplayPercents` for 2-item complement |
| `CANSummaryCards` | Same 2-item complement fix |
| `PortfolioSummaryCards.helpers.js` | Single normalisation pass across all portfolios |
| `PortfolioLegend` | Removed local float guard; renders pre-normalised `percent` directly |
| `HorizontalStackedBar` | Filters/sizes by `value` (not `percent` string); a11y labels speak guarded display percent |
| `LineGraph` + `ReverseLineGraph` | CSS width from `value/total`; 2% minimum floor for tiny bars |
| `ResponsiveDonutWithInnerPercent` | `applyMinimumArcValue` now applied internally (closes #5474) |
| `ProjectTypeSummaryCard` | Switched from `computeDisplayPercent` (singular) to `computeDisplayPercents` (plural); removed local `applyMinimumArcValue` workaround |
| `ReceivedFundingCard` | Removed vestigial `calculatePercent` call and dead `percent` fields |

**Dead code removed:** `PortfolioFundingByCAN` directory — component was unreachable (zero imports) and referenced Redux slices that were never registered in the store.

**Dev review page:** `/dev/data-viz` route (DEV-only, gated by `import.meta.env.DEV`) with worst-case stub data for each affected component, showing green `✓ FIXED` labels.

## Issue

Closes #5513
Closes #5474

## How to test

1. Start the dev stack: `docker compose up --build`
2. Navigate to `http://localhost:3000/dev/data-viz`
3. All sections should show green `✓ FIXED` labels — no red `✗ BUG` labels visible
4. Verify section 1a (`BLIStatusSummaryCard`): dominant Draft shows `>99%`, peers show `<1%`, tiny arc is visible
5. Verify section 1c: the 0.5% Planned arc is visible as a thin slice in the donut
6. Verify section 2a/2b (`AgreementSpendingSummaryCard`): same `>99%`/`<1%` behaviour
7. Verify section 3a/3b (`AgreementSpendingCards`): stacked bar renders correctly, no missing bars
8. Verify section 4a (`PortfolioFunding`): carry-forward bar visually dominates; legend shows `>99%` and `<1%`
9. Verify section 5a (`PortfolioSummaryCards`): CC legend shows `>99%`, not `100%`
10. Run unit tests: `cd frontend && bun run test --watch=false` — 254 files, 0 failures

## A11y impact

- [x] Accessibility changes included and validated against WCAG 2.1 AA intent

`HorizontalStackedBar` `aria-label` and `sr-only` text now speak the guarded display percent (e.g. `">99% of budget"` instead of `"100% of budget"` — the former is accurate, the latter is contradicted by the visible `<1%` peer). Screen readers previously announced contradictory information; this PR corrects that.

## Screenshots
SCR-20260415-nxfb

SCR-20260415-odij

SCR-20260415-oeth

SCR-20260415-onql

SCR-20260415-oocm

SCR-20260415-pile

SCR-20260415-otzz

image

See `/dev/data-viz` review page for visual before/after comparison of all affected components with worst-case stub data.

## Definition of Done Checklist
- [x] OESA: Code refactored for clarity
- [x] OESA: Dependency rules followed
- [x] Automated unit tests updated and passed
- [x] Automated integration tests updated and passed
- [x] Automated quality tests updated and passed
- [x] Automated load tests updated and passed
- [x] Automated a11y tests updated and passed
- [x] Automated security tests updated and passed
- [x] 90%+ Code coverage achieved
- [-] Form validations updated

## Links

- Issue #5513
- Issue #5474

### CODE REVIEW

File: frontend/src/components/UI/Cards/BudgetCard/ReceivedFundingCard.test.jsx [hunk 1]
```diff
@@ -66,7 +66,11 @@ describe("ReceivedFundingCard", () => {
                 totalFunding={5000}
             />
         );
-        expect(screen.getByTestId("line-graph-left-bar")).toHaveStyle("flex: 0 1 20%");
+        // Width is derived from value proportion: 1000 / (1000 + 5000) = 16.67%
+        // (previously was 20% because it used the display percent field directly)
+        expect(screen.getByTestId("line-graph-left-bar")).toHaveStyle(
+            "flex: 0 1 16.666666666666664%"
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/components/UI/Cards/BudgetCard/ReceivedFundingCard.test.jsx]: Fixed. Right `value` is now `Math.max((totalFunding ?? 0) - (totalReceived ?? 0), 0)`. Width = received / totalFunding = 20%. Test updated to assert `flex: 0 1 20%`.

File: frontend/src/components/Portfolios/PortfolioSummaryCards/PortfolioSummaryCards.helpers.js [hunk 2]
```diff
@@ -112,22 +112,28 @@ export const transformPortfoliosToChartData = (sortedPortfolios, totalBudget) =>
...
-            color: FALLBACK_COLOR,
-            percent
+            color: FALLBACK_COLOR
         });
     });
 
+    // Flatten all real items, apply cross-item percent normalisation in one
+    // pass (>99% cap + <1% guard), then re-index by id for fast lookup.
+    const allRealItems = columns.flat();
+    const normalised = computeDisplayPercents(allRealItems);
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/components/Portfolios/PortfolioSummaryCards/PortfolioSummaryCards.helpers.js]: Fixed. When unknown portfolios are truncated from the grid, a synthetic remainder item (`__truncated-remainder__`) is injected before calling `computeDisplayPercents` so the true `totalBudget` is used as the denominator. The `eslint-disable` comment is removed now that `totalBudget` is referenced. New test covers the truncated-items case.

File: frontend/src/helpers/utils.js [hunk 3]
```diff
@@ -565,3 +565,112 @@ export function convertToCurrency(value) {
...
+ *   present but too small to show as 1%).
+ * - Otherwise returns the rounded integer percent.
+ *
+ * @param {number} value - The slice value.
+ * @param {number} total - The sum across all slices.
+ * @returns {number|string} Rounded integer percent, "<1", or 0.
+ */
+export const computeDisplayPercent = (value, total) => {
+    if (total === 0 || value === 0) return 0;
+    const exact = (value / total) * 100;
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/helpers/utils.js]: Fixed. Both `value` and `total` are now coerced via `Number()` with an `isFinite` guard before any arithmetic, so numeric strings and `NaN`/`Infinity` inputs return `0` correctly. Tests added for string coercion and non-finite inputs.

File: frontend/src/helpers/utils.js [hunk 4]
```diff
@@ -565,3 +565,112 @@ export function convertToCurrency(value) {
...
+    const total = items.reduce((sum, item) => sum + (item.value ?? 0), 0);
+
+    if (total === 0) {
+        return items.map((item) => ({ ...item, percent: 0 }));
+    }
+
+    const nonZeroCount = items.filter((item) => (item.value ?? 0) > 0).length;
+
+    return items.map((item) => {
+        const value = item.value ?? 0;
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/helpers/utils.js]: Fixed. Introduced a `toFiniteNumber` helper inside the function that calls `Number(v ?? 0)` and guards against non-finite results. Applied to the total reduction and all `item.value` accesses so numeric strings never trigger string concatenation.

File: frontend/src/helpers/utils.js [hunk 5]
```diff
@@ -565,3 +565,112 @@ export function convertToCurrency(value) {
...
+    // Floor any non-zero slice that is below the minimum
+    const adjustedItems = items.map((item) => ({
+        ...item,
+        value: item.value > 0 && item.value < minValue ? minValue : item.value
+    }));
+
+    // How much was added in total by flooring
+    const addedValue = adjustedItems.reduce((sum, item, index) => sum + (item.value - items[index].value), 0);
+
+    if (addedValue <= 0) return adjustedItems;
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/helpers/utils.js]: Fixed. Added an early `needsAdjustment` check before allocating `adjustedItems`. If no slice requires flooring, the original `items` reference is returned immediately with no allocation.

File: frontend/src/components/UI/DataViz/LineGraph/LineGraph.jsx [hunk 6]
```diff
@@ -18,19 +18,48 @@ import styles from "./styles.module.scss";
...
+
+    // Apply a minimum visible floor of 2% for non-zero bars so a tiny sliver
+    // is never invisible. Cap the opposing bar so the two widths stay coherent.
+    const MIN_WIDTH = 2;
+    let leftFlexWidth = rawLeftWidth;
+    if (rawLeftWidth > 0 && rawLeftWidth < MIN_WIDTH) {
+        leftFlexWidth = MIN_WIDTH;
+    } else if (rawRightWidth > 0 && rawRightWidth < MIN_WIDTH) {
+        leftFlexWidth = 100 - MIN_WIDTH;
+    }
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/components/UI/DataViz/LineGraph/LineGraph.jsx]: Fixed. Width logic is now backwards-compatible: if `leftPercent` is a finite number, it is used directly for `flex-basis` (preserving correct proportions for callers like `BudgetCard`/`CanCard` that pass `totalFunding` as the right value). The value-proportional fallback (`left/(left+right)`) is used only when `leftPercent` is a display string (">99"/"<1"), which applies only to the `PortfolioFunding`/`CANSummaryCards` callers that pass complementary values. The shared `resolveLeftFlexWidth` helper is extracted to `util.js` and reused by both `LineGraph` and `ReverseLineGraph`.

File: frontend/src/components/UI/DataViz/LineGraph/ReverseLineGraph.jsx [hunk 7]
```diff
@@ -25,30 +16,58 @@ import styles from "./styles.module.scss";
...
+
+    // Apply a minimum visible floor of 2% for non-zero bars so a tiny sliver
+    // is never invisible. Cap the opposing bar so the two widths stay coherent.
+    const MIN_WIDTH = 2;
+    let leftFlexWidth = rawLeftWidth;
+    if (rawLeftWidth > 0 && rawLeftWidth < MIN_WIDTH) {
+        leftFlexWidth = MIN_WIDTH;
+    } else if (rawRightWidth > 0 && rawRightWidth < MIN_WIDTH) {
+        leftFlexWidth = 100 - MIN_WIDTH;
+    }
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/components/UI/DataViz/LineGraph/ReverseLineGraph.jsx]: Fixed. Same `resolveLeftFlexWidth` strategy applied — uses numeric `percent` when available, falls back to value-proportional only for display strings. Both components now import from `util.js`.

File: frontend/src/components/UI/Cards/BudgetCard/ReceivedFundingCard.jsx [hunk 8]
```diff
@@ -20,19 +19,16 @@ import { calculatePercent } from "../../../../helpers/utils";
...
         {
             id: 1,
             value: totalReceived,
-            color: "var(--data-viz-budget-graph-1)",
-            percent: receivedPercent
+            color: "var(--data-viz-budget-graph-1)"
         },
         {
             id: 2,
             value: totalFunding,
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/components/UI/Cards/BudgetCard/ReceivedFundingCard.jsx]: Fixed. Right `value` is now `Math.max((totalFunding ?? 0) - (totalReceived ?? 0), 0)` — the complement — so `left + right = totalFunding` and the bar width is correctly `received / totalFunding`.

File: frontend/cypress/e2e/agreementList.cy.js [hunk 9]
```diff
@@ -513,7 +513,10 @@
                 let totalPercent = 0;
                 $tags.each((_, tag) => {
                     const percentText = Cypress.$(tag).text();
-                    totalPercent += parseInt(percentText, 10);
+                    // Strip display-string prefixes (">99%" → 99, "<1%" → 1, "0%" → 0)
+                    // so the sum check works correctly after cross-item normalisation.
+                    const normalized = percentText.replace(/^[><]/, "").replace("%", "").trim();
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/cypress/e2e/agreementList.cy.js]: Fixed in d11456535. Changed `.replace("%", "")` to `.replace(/%/g, "")` — using a regex with the global flag makes intent explicit and resolves the incomplete-sanitization finding. Note: this is test code parsing a known display string, not user-input sanitization, so the risk was already zero — but the regex form is unambiguously correct.

**Review by johndeange — APPROVED** [review]
## Code Review: Percentage Label Consistency Fix

### Overview

Well-structured PR that addresses a codebase-wide class of rounding and display-consistency bugs across all data-viz components. The centralization of rounding logic into three shared utilities (`computeDisplayPercent`, `computeDisplayPercents`, `applyMinimumArcValue`) is the right architectural choice and will prevent an entire class of bugs going forward.

---

### Issues (minor, non-blocking)

**1. PR description says `>99%` but implementation shows `99%`**

Steps 4, 8, and 9 in "How to test" reference `>99%`, but the code and tests correctly use plain `99%` per the Figma spec. The AGENTS.md documentation is correct — just a PR description inconsistency.

**2. `LineGraph.jsx` / `ReverseLineGraph.jsx` — tautological fullness check**

The fallback branch reads:
```js
(leftValue ?? 0) >= (leftValue ?? 0) + (rightValue ?? 0)
```
This simplifies to `rightValue <= 0`. Consider clarifying in both files (4 occurrences total).

**3. No unit tests for `resolveLeftFlexWidth`**

`frontend/src/components/UI/DataViz/LineGraph/util.js` has non-trivial branching (numeric vs. string percent, min-width floors, zero total) but no dedicated tests. Consider adding direct coverage for the key branches.

---

### Suggestions

- **`computeDisplayPercents` edge case**: When items include `"<1"` strings, the largest-remainder pass is skipped, so numeric labels may not sum to exactly 100 in mixed cases. A JSDoc note on this intentional tradeoff would help future contributors.
- **Stale doc reference**: `docs/developers/frontend/accessibility/issue-5149-wave-3.md:89` still references the deleted `PortfolioFundingByCAN` component.

---

### Positive Observations

- Excellent centralization — 12+ independent `calculatePercent` calls replaced by a single `computeDisplayPercents` pass
- Thorough test coverage with edge cases (dominant, equal split, single item, tiny arc, zero total)
- Comprehensive AGENTS.md documentation with clear do/don't examples
- Good dead code removal (`PortfolioFundingByCAN`)
- A11y improvement — `HorizontalStackedBar` aria-labels now speak the guarded percent
- Moving `applyMinimumArcValue` into `ResponsiveDonutWithInnerPercent` eliminates consumer-side mistakes
- DEV debug page (`/dev/data-viz`) is a thoughtful addition for manual verification

LGTM — approve with minor suggestions above.

**Comment by fpigeonjr** [comment]
Addressed all items from @johndeange's review in commit 1f4ee7d86:

**Issue 1 – PR description `>99%` vs `99%`**: No code change needed; this was only a prose inconsistency in the PR description. The code and AGENTS.md are correct.

**Issue 2 – Tautological fullness check** (`LineGraph.jsx` / `ReverseLineGraph.jsx`): The expression `(leftValue ?? 0) >= (leftValue ?? 0) + (rightValue ?? 0)` was replaced with the clearer `(rightValue ?? 0) <= 0` (and the symmetric `(leftValue ?? 0) <= 0` for the right side). Same fix applied in both files.

**Issue 3 – No unit tests for `resolveLeftFlexWidth`**: Added direct coverage in `util.test.js` for all key branches: numeric percent (normal, zero, tiny <2%, dominant ≥98%), and the string/display-percent fallback ("<1", normal complementary values, zero total, dominant with string percent).

**Suggestion – `computeDisplayPercents` JSDoc**: Added a note to the JSDoc explaining the intentional tradeoff: when any item produces a string percent (`"<1"` or capped `99`), the largest-remainder integer-sum pass is skipped for the whole set.

**Stale doc reference** – `docs/developers/frontend/accessibility/issue-5149-wave-3.md:89`: Updated the `PortfolioFundingByCAN` reference to note that the component was subsequently removed in this PR (#5515).

All 2758 tests pass and lint is clean.

**Review by weimiao67 — APPROVED** [review]
